taichi/program/callable.h

```diff
@@ -16,22 +16,45 @@ class TI_DLL_EXPORT Callable {
   std::unique_ptr<FrontendContext> context{nullptr};
 
   struct Arg {
-    DataType dt;
     bool is_array{
         false};  // This is true for both ndarray and external array args.
-    std::size_t total_dim{0};             // total dim of array
-    std::vector<int> element_shape = {};  // shape of each element
+    std::size_t total_dim{0};  // total dim of array
 
     explicit Arg(const DataType &dt = PrimitiveType::unknown,
                  bool is_array = false,
                  std::size_t size_unused = 0,
                  int total_dim = 0,
-                 std::vector<int> element_shape = {})
-        : dt(dt),
-          is_array(is_array),
-          total_dim(total_dim),
-          element_shape(std::move(element_shape)) {
+                 std::vector<int> element_shape = {}) {
+      if (dt->is<PrimitiveType>() && element_shape.size() > 0) {
```

Comment:
Could you explain this case? Why is it possible that the user declares a value with some primitive type but has non-zero shape? I thought we handle all the cases where `element_shape.size() > 0` at the frontend. Thanks!

Reply:
Good point, thanks! 

In general it is a historical issue. Taichi used to represent TensorType with a combination of "PrimitiveType" & "element_shape" and there are a bunch of interfaces designed like this (it allows the users to create TensorType by passing in PrimitiveType + element_shape)

In this refactor PR, we mainly wanna remove the "element_shape" member in the objects (class Arg, class ExternalTensorExpression, ...), and force them to use TensorType in thier "dtype" member. However we still wanna keep the interfaces unchanged so as to minimize possible regressions.

Reply:
Got it. Makes sense to me. Thanks for the explanation!